### DISCUSSION TITLE
Is there a way to bypass the limit of 10,000 items?

### DISCUSSION THREAD

**Discussion by glaroc** [stac-utils/stac-fastapi#737] [Q&A] [answered]
We have multiple use cases where the only differences between the items in a collection are in a custom property variable. For example, we have models outputs for 5k species for multiple years and multiple models. So, it often happens that we need to filter the catalog for all results for a given species, as specified by a variable in the item's properties with the species name, or for all results from a particular model. 

For this reason, the limit of 10,000 items returned by the collection/{}/items or /search endpoints is really problematic for us. Is there any way to bypass this or simply to eliminate the limit in Stac-fastapi ?

**Comment by vincentsarago** [comment]
you can set a custom Search model with a max limit >1000

https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/search.py#L184-L190

```python
from pydantic import Field
from stac_pydantic.api import Search

class BaseSearchPostRequest(Search):
    """Base arguments for POST Request."""

    limit: Optional[int] = Field(
        10,
        description="Limits the number of results that are included in each page of the response.",  # noqa: E501
    )


api = StacApi(
    search_post_request_model=BaseSearchPostRequest
)
```

I don't think there is a upper limit for the /items endpoint

**Comment by glaroc** [comment]
Thank you, we'll give this a try. 

> I don't think there is a upper limit for the /items endpoint

I believe there is. We can't ever seem to get more than 10,000 items, regardless of the limit= parameter.

**Reply by vincentsarago** [reply]
which backend are you using? 

if you're using elastic search backend I think this won't be configurable https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/blob/f2665ca8317cf1c589ce2368bda8e5174d82ee77/stac_fastapi/opensearch/stac_fastapi/opensearch/database_logic.py#L619-L630 

you can open an issue in the backend repo directly

**Reply by glaroc** [reply]
We are using pgstac.

**Reply by jonhealy1** [reply]
Pagination?

**Reply by jonhealy1** [reply]
Elasticsearch itself has a built-in 10,000 item return limit so switching over would not help ...

**Reply by glaroc** [reply]
Is pagination available with stac-fastapi ? I haven't seen any docs on this.

**Reply by geospatial-jeff** [reply]
Item-level pagination is available for most backends (it is a requirement of OGC API Features).  Pgstac does implement it, look for a `next` and `previous` rel-type in the top-level links returned by the various search endpoints.

You will quickly run into performance bottlenecks using such a large limit, pagination is absolutely the way to go here.

**Reply by vincentsarago** [reply]
the limit in the `/items` endpoint comes from https://github.com/stac-utils/stac-fastapi/blob/4df4947b9822271b0912abe766234edb26a78a59/stac_fastapi/api/stac_fastapi/api/models.py#L111-L120

so I don't understand how the limit is max out to 10000 🤷 maybe it's a pgstac thing, cc @bitner 🤷

**Comment by vincentsarago** [comment]
@glaroc which version of stac-fastapi are you using?

**Comment by bitner** [comment, marked as answer]
There is no max limit set in pgstac.

The max limit of 10,000 comes from [here](https://github.com/stac-utils/stac-fastapi/blob/4df4947b9822271b0912abe766234edb26a78a59/stac_fastapi/types/stac_fastapi/types/search.py#L18 ). 

Unconstrained large limits run the risk of causing memory issues or timeouts.

Pagination is definitely the preferred method to get more results than that. If you use pystac-client to interact with a pagination enabled stac api, the library completely takes care of the pagination for you and running an search returns an iterator that will transparently make all the paginated requests for you.

**Reply by vincentsarago** [reply]
@bitner I don't see this used in the `/items` endpoint

**Reply by vincentsarago** [reply]
but it will with #738
